## Data Collection

In [1]:
# importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#listings data obtained from http://insideairbnb.com/get-the-data
listings = "./data/listings.csv" 

df = pd.read_csv(listings, low_memory=False)
df.shape


(41533, 18)

In [3]:
# Inspecting the data
df.head()

id                                      name  host_id  host_name  \
0  5136  Spacious Brooklyn Duplex, Patio + Garden     7378    Rebecca   
1  5203        Cozy Clean Guest Room - Family Apt     7490  MaryEllen   
2  5121                           BlissArtsSpace!     7356      Garon   
3  5178          Large Furnished Room Near B'way　     8967   Shunichi   
4  2595                     Skylit Midtown Castle     2845   Jennifer   

  neighbourhood_group       neighbourhood  latitude  longitude  \
0            Brooklyn         Sunset Park  40.66265  -73.99454   
1           Manhattan     Upper West Side  40.80380  -73.96751   
2            Brooklyn  Bedford-Stuyvesant  40.68535  -73.95512   
3           Manhattan             Midtown  40.76457  -73.98317   
4           Manhattan             Midtown  40.75356  -73.98559   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt    275              21                  3  2022-08-10   
1     Private room     75               2                118  2017-07-21   
2     Private room     60              30                 50  2019-12-02   
3     Private room     68               2                559  2022-11-20   
4  Entire home/apt    175              30                 49  2022-06-21   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.03                               1               267   
1               0.73                               1                 0   
2               0.30                               2               322   
3               3.38                               1                79   
4               0.31                               3               365   

   number_of_reviews_ltm license  
0                      1     NaN  
1                      0     NaN  
2                      0     NaN  
3                     50     NaN  
4                      1     NaN

In [4]:
# importing detailed listing info for reference
df_detail = pd.read_csv("./data/listings_detail.csv")



C:\Users\Nizar\AppData\Local\Temp\ipykernel_20848\1395750445.py:2: DtypeWarning: Columns (68) have mixed types. Specify dtype option on import or set low_memory=False.
  df_detail = pd.read_csv("./data/listings_detail.csv")


## Data Definition

In [5]:
# Generating a df of label descriptions
labels_temp = pd.read_excel("./data/Inside Airbnb Data Dictionary.xlsx")
labels_clean = labels_temp.iloc[6:]
labels_clean.columns = labels_clean.iloc[0]
labels = labels_clean.iloc[1:]
labels.head()

6          Field      Type Calculated  \
7             id   integer        NaN   
8    listing_url      text          y   
9      scrape_id    bigint          y   
10  last_scraped  datetime          y   
11        source      text        NaN   

6                                         Description Reference  
7          Airbnb's unique identifier for the listing       NaN  
8                                                 NaN       NaN  
9             Inside Airbnb "Scrape" this was part of       NaN  
10  UTC. The date and time this listing was "scrap...       NaN  
11  One of "neighbourhood search" or "previous scr...       NaN

In [6]:
# dropping unnecessary features
columns_to_drop = ['Calculated', 'Reference']
labels = labels.drop(labels=columns_to_drop, axis=1)
labels = labels.reset_index(drop=True).rename_axis(None, axis=1)

In [7]:
# filtering for columns of interest
features = list(df.columns)
labels.loc[labels['Field'].isin(features)]

Field      Type  \
0                               id   integer   
5                             name      text   
9                          host_id   integer   
11                       host_name      text   
27                   neighbourhood      text   
30                        latitude   numeric   
31                       longitude   numeric   
33                       room_type      text   
40                           price  currency   
41                  minimum_nights   integer   
54                availability_365   integer   
56               number_of_reviews   integer   
57           number_of_reviews_ltm   integer   
60                     last_review      date   
68                         license      text   
70  calculated_host_listings_count   integer   
74               reviews_per_month   numeric   

                                          Description  
0          Airbnb's unique identifier for the listing  
5                                 Name of the listing  
9        Airbnb's unique identifier for the host/user  
11  Name of the host. Usually just the first name(s).  
27                                                NaN  
30  Uses the World Geodetic System (WGS84) project...  
31  Uses the World Geodetic System (WGS84) project...  
33  [Entire home/apt|Private room|Shared room|Hote...  
40                      daily price in local currency  
41  minimum number of night stay for the listing (...  
54  avaliability_x. The availability of the listin...  
56              The number of reviews the listing has  
57  The number of reviews the listing has (in the ...  
60                 The date of the last/newest review  
68             The licence/permit/registration number  
70  The number of listings the host has in the cur...  
74  The number of reviews the listing has over the...

In [8]:
# Ensure the Dtypes match up
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41533 entries, 0 to 41532
Data columns (total 18 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              41533 non-null  int64  
 1   name                            41520 non-null  object 
 2   host_id                         41533 non-null  int64  
 3   host_name                       41528 non-null  object 
 4   neighbourhood_group             41533 non-null  object 
 5   neighbourhood                   41533 non-null  object 
 6   latitude                        41533 non-null  float64
 7   longitude                       41533 non-null  float64
 8   room_type                       41533 non-null  object 
 9   price                           41533 non-null  int64  
 10  minimum_nights                  41533 non-null  int64  
 11  number_of_reviews               41533 non-null  int64  
 12  last_review                     

## Data Cleaning



Ill start by inspecting the nulls.

In [9]:
df.isnull().sum().sort_values(ascending=False)

license                           41532
reviews_per_month                  9393
last_review                        9393
name                                 13
host_name                             5
minimum_nights                        0
number_of_reviews_ltm                 0
availability_365                      0
calculated_host_listings_count        0
number_of_reviews                     0
id                                    0
room_type                             0
longitude                             0
latitude                              0
neighbourhood                         0
neighbourhood_group                   0
host_id                               0
price                                 0
dtype: int64

The licence information is missing and has no effect on the project so it can be dropped.

There are 13 entries under the `name` field that are missing, as well as 9393 entries related to reviews. Lets insepct these to see whats missing.

In [10]:
df = df.drop('license', axis=1)

In [11]:
df[df.name.isna()]

id name    host_id host_name neighbourhood_group  \
1827    2232600  NaN   11395220      Anna           Manhattan   
2831    4209595  NaN   20700823     Jesse           Manhattan   
2982    4370230  NaN   22686810   Michaël           Manhattan   
3099    4581788  NaN   21600904     Lucie            Brooklyn   
3296    4774658  NaN   24625694      Josh           Manhattan   
4395    6782407  NaN   31147528  Huei-Yin            Brooklyn   
5916    9325951  NaN   33377685  Jonathan           Manhattan   
6343    9787590  NaN   50448556    Miguel           Manhattan   
6675   10116081  NaN   51913270    Andrew           Manhattan   
6711   10052289  NaN   49522403   Vanessa            Brooklyn   
7816   12988898  NaN   71552588    Andrea               Bronx   
7839   12797684  NaN   69715276       Yan           Manhattan   
11742  21044649  NaN  151547086     Tanya               Bronx   

            neighbourhood  latitude  longitude        room_type  price  \
1827         East Village  40.73192  -73.98819  Entire home/apt    700   
2831    Greenwich Village  40.73323  -73.99294  Entire home/apt    225   
2982               Nolita  40.72100  -73.99536  Entire home/apt    215   
3099         Williamsburg  40.71370  -73.94378     Private room    150   
3296   Washington Heights  40.85111  -73.93009     Private room     40   
4395         Williamsburg  40.71354  -73.93882     Private room     45   
5916       Hell's Kitchen  40.76617  -73.98435  Entire home/apt    190   
6343               Harlem  40.80551  -73.95069  Entire home/apt    300   
6675              Midtown  40.75939  -73.96949  Entire home/apt    200   
6711          Brownsville  40.66409  -73.92314     Private room     80   
7816              Fordham  40.86046  -73.88378      Shared room    130   
7839      Upper West Side  40.79843  -73.96404     Private room    100   
11742             Norwood  40.87553  -73.88601     Private room     10   

       minimum_nights  number_of_reviews last_review  reviews_per_month  \
1827               60                 28  2015-06-08               0.27   
2831               30                  1  2015-01-01               0.01   
2982               30                  5  2016-01-02               0.05   
3099               30                  0         NaN                NaN   
3296               30                  0         NaN                NaN   
4395               30                  0         NaN                NaN   
5916               30                  1  2016-01-05               0.01   
6343               30                  0         NaN                NaN   
6675               30                  0         NaN                NaN   
6711               30                  3  2016-08-18               0.04   
7816               30                  0         NaN                NaN   
7839               30                  0         NaN                NaN   
11742               3                 24  2021-03-25               0.38   

       calculated_host_listings_count  availability_365  number_of_reviews_ltm  
1827                                1               359                      0  
2831                                1                 0                      0  
2982                                1                 0                      0  
3099                                1                 0                      0  
3296                                1                 0                      0  
4395                                1                 0                      0  
5916                                1                 0                      0  
6343                                5                 0                      0  
6675                                1                 0                      0  
6711                                1                 0                      0  
7816                                1                 0                      0  
7839                                2     

There doesn't appear to be any visual correlation between these listings in terms of location or host. Most of them have no reviews, and without a description highlighting amenities, we wont be able to include them in the comparisons. For now, I will name them 'Unnamed Listing' and we'll decide if any value can be extracted during the EDA step.

In [12]:
df.name.fillna('Unnamed Listing', inplace=True)

In [13]:
df.price.describe()

count    41533.000000
mean       221.978282
std        919.502236
min          0.000000
25%         80.000000
50%        131.000000
75%        220.000000
max      98159.000000
Name: price, dtype: float64

In [14]:
# The minimum value indicates there are listings for $0. 
df[df.price == 0]

id                                  name    host_id  \
20487  40560656  The Hoxton, Williamsburg - Cosy Room  273324213   
21001  41792753                 Mint House At 70 Pine  197053492   
21010  41740615            The James New York - NoMad  268417148   
21011  41740622       Gardens Suites Hotel by Affinia  269311462   
21148  42065543                  Broadway Plaza Hotel  307634016   
21152  42065545                       Carvi Hotel NYC  310429455   
21154  42065547                          Hotel Five44  308721299   
21155  42065555                          Pod Hotel BK  309714886   
21162  42065562                         Artezen Hotel  307633956   
21167  42065563                     Opera House Hotel  309772430   
21168  42065564                   The Wall Street Inn  314151200   
21224  42228997                  Sister City New York  314197504   
21278  42279171                        Leon Hotel LES  265458818   
21335  42384501                          Hotel Hayden  262458398   
21678  43035720     Executive Hotel LeSoleil New York  318559292   
21787  43078550                     The Lex Hotel NYC  334334264   
21788  43078552                             SIXTY LES  342053968   
21900  43205598                      Wellington Hotel  335389657   
22006  43247386         6 Columbus Central Park Hotel  335072254   
22010  43247472                              Hotel 57  324955773   
22014  43247631                       Belvedere Hotel  318788301   
22536  44567521            The Standard, East Village  360662584   
23282  46059074  CITIZENM NEW YORK TIMES SQUARE HOTEL  373324108   
23356  46087899                      The William Vale  373522899   
23465  46336133                Equinox Hotel New York  375044940   
23466  46251446        citizenM New York Bowery hotel  374516933   
23742  46723973                          Ink 48 Hotel  376877842   
24518  48089897                          Condor Hotel  261016212   
24785  48417136                          HGU New York  390810530   
24795  48325676                          Hotel Beacon  390077597   

                              host_name neighbourhood_group  \
20487           The Hoxton Williamsburg            Brooklyn   
21001             Mint House At 70 Pine           Manhattan   
21010                   The James NoMad           Manhattan   
21011   Gardens Suites Hotel By Affinia           Manhattan   
21148                    Broadway Plaza           Manhattan   
21152                       Carvi Hotel           Manhattan   
21154                      Hotel Five44           Manhattan   
21155                         Pod Hotel            Brooklyn   
21162                     Artezen Hotel           Manhattan   
21167                 Opera House Hotel               Bronx   
21168               The Wall Street Inn           Manhattan   
21224                       Sister City           Manhattan   
21278                    Leon Hotel LES           Manhattan   
21335                      Hotel Hayden           Manhattan   
21678                Executive LeSoleil           Manhattan   
21787                         Lex Hotel           Manhattan   
21788                         Sixty Les           Manhattan   
21900                  Wellington Hotel           Manhattan   
22006  Six Columbus  Central Park Hotel           Manhattan   
22010                          Hotel 57           Manhattan   
22014                              Beth           Manhattan   
22536                      The Standard           Manhattan   
23282                          CitizenM           Manhattan   
23356                  The William Vale            Brooklyn   
23465                     Equinox Hotel           Manhattan   
23466                          CitizenM           Manhattan   
23742                       Ink48 Hotel           Manhattan   
24518                  The Condor Hotel            Brooklyn   
24785                            Marcos           Manhattan   
24795         

It looks like there are 30 listings with a price of $0. This may be true or it may be an error, but there is no efficient way to check. Additionally, these entries are all listed as Hotel room for the `room_type`. Since the start-up will not be including these in their listing, we may drop them all together. For now, I will drop these entries since they wont help with the price.

In [15]:
df = df[df.price != 0]
df.price.describe()

count    41503.000000
mean       222.138737
std        919.815135
min         10.000000
25%         80.000000
50%        131.000000
75%        220.000000
max      98159.000000
Name: price, dtype: float64

Thats better. Finally, to adress the reviews. I will replace the null values in the `reviews_per_month` column with 0. As for the `last_review` column, it may be of help later in measuring the activity of listings but it will be dropped for now.

In [16]:
df.reviews_per_month.fillna(0)
df = df.drop('last_review', axis=1)
df.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'reviews_per_month',
       'calculated_host_listings_count', 'availability_365',
       'number_of_reviews_ltm'],
      dtype='object')

`id` and `name` are the only identifiers needed. The `host_name` will be dropped for privacy reasons, but I will keep the `host_id` for now.

In [17]:
df_final = df.drop('host_name', axis=1)
print(df_final.columns)

Index(['id', 'name', 'host_id', 'neighbourhood_group', 'neighbourhood',
       'latitude', 'longitude', 'room_type', 'price', 'minimum_nights',
       'number_of_reviews', 'reviews_per_month',
       'calculated_host_listings_count', 'availability_365',
       'number_of_reviews_ltm'],
      dtype='object')


The features that I plan to investigate their effect on price are (location, property type, amenities, number of bedrooms and baths, local market conditions, and past rental data).

For location: I'll start off by using the `neighborhood` and `neighbourhood_group` as classifiers. The geocode location can be added later for a more in depth analysis if needed.

The `room_type` entry comprises of the following unique values: 'Entire home/apt', 'Private room', 'Hotel room', 'Shared room'. Hotel room entries may be dropped in the future.

df_detail contains `amenities` which is a json object of self-reported available amenities. This will remain in a seperate dataframe for now.

`accommodates` and `bedrooms` are integers relating to number of guests.
`bathrooms_text` contains string data about the number of baths.
These will also be added in the EDA step.

local market conditions will comprise of the hosts mentioned `availability_365` as well as the `calculated_host_listings_count`. 

Past rental data in this case will include: the `number_of_reviews and `number_of_reviews_ltm` which is specific to the past 12 months.

The remaining features will be left in during the EDA step to check for other correlations. 


In [18]:
listings_final = df_final.merge(df_detail[['accommodates', 'bedrooms', 'bathrooms_text', 'amenities', 'id']], on='id', how='left')

In [19]:
df_final.shape, listings_final.shape

((41503, 15), (41503, 19))

In [20]:
listings_final.head()

id                                      name  host_id  \
0  5136  Spacious Brooklyn Duplex, Patio + Garden     7378   
1  5203        Cozy Clean Guest Room - Family Apt     7490   
2  5121                           BlissArtsSpace!     7356   
3  5178          Large Furnished Room Near B'way　     8967   
4  2595                     Skylit Midtown Castle     2845   

  neighbourhood_group       neighbourhood  latitude  longitude  \
0            Brooklyn         Sunset Park  40.66265  -73.99454   
1           Manhattan     Upper West Side  40.80380  -73.96751   
2            Brooklyn  Bedford-Stuyvesant  40.68535  -73.95512   
3           Manhattan             Midtown  40.76457  -73.98317   
4           Manhattan             Midtown  40.75356  -73.98559   

         room_type  price  minimum_nights  number_of_reviews  \
0  Entire home/apt    275              21                  3   
1     Private room     75               2                118   
2     Private room     60              30                 50   
3     Private room     68               2                559   
4  Entire home/apt    175              30                 49   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.03                               1               267   
1               0.73                               1                 0   
2               0.30                               2               322   
3               3.38                               1                79   
4               0.31                               3               365   

   number_of_reviews_ltm  accommodates  bedrooms bathrooms_text  \
0                      1             4       2.0      1.5 baths   
1                      0             1       1.0  1 shared bath   
2                      0             2       1.0            NaN   
3                     50             2       1.0         1 bath   
4                      1             1       NaN         1 bath   

                                           amenities  
0  ["HDTV with Disney+, HBO Max, standard cable, ...  
1  ["Hair dryer", "Smoke alarm", "Carbon monoxide...  
2  ["Heating", "Wifi", "Long term stays allowed",...  
3  ["Shampoo", "Conditioner", "TV", "Drying rack ...  
4  ["Carbon monoxide alarm", "Stove", "Bathtub", ...

In [21]:
listings_final.to_csv('listings_wrangled', index=False)